### PR TITLE
Pre-compute `LocalDefId` <-> `HirId` mappings and remove `NodeId` <-> `HirId` conversion APIs

### PR DESCRIPTION
**Pull request by marmeladema**
cc #50928 

I don't know who is exactly the best person to review this.

r? @petrochenkov

### CODE REVIEW

**Comment by marmeladema** [comment]
Next step is probably to move
```Rust
    node_id_to_def_id: FxHashMap<ast::NodeId, LocalDefId>,
    def_id_to_node_id: IndexVec<LocalDefId, ast::NodeId>,
```
mappings from `hir::Definitions` to `Resolver` such that any references to `NodeId` are gone with it.

**Comment by petrochenkov** [comment]
What is the end goal here, removing any mentions of `NodeId` from `struct Definitions`?

`Definitions` is a pretty heterogeneous structure.
Perhaps it can be split into two structures instead, one for fields used before lowering to HIR, and one for fields used after lowering to HIR?
```rust
struct Definitions {
    // Both before and after (or filled before, used after)
    table: DefPathTable,
    parent_modules_of_macro_defs: FxHashMap<ExpnId, DefId>,
    expansions_that_defined: FxHashMap<LocalDefId, ExpnId>,

    // Before
    def_id_to_span: IndexVec<LocalDefId, Span>,
    node_id_to_def_id: FxHashMap<ast::NodeId, LocalDefId>,
    def_id_to_node_id: IndexVec<LocalDefId, ast::NodeId>,
    next_disambiguator: FxHashMap<(LocalDefId, DefPathData), u32>,
    invocation_parents: FxHashMap<ExpnId, LocalDefId>,
    placeholder_field_indices: FxHashMap<ast::NodeId, usize>,

    // ???
    node_id_to_hir_id: IndexVec<ast::NodeId, Option<hir::HirId>>,
    hir_id_to_node_id: FxHashMap<hir::HirId, ast::NodeId>,
}
```

Having the "before" fields in the `ResolverOutputs` version of `Definitions` doesn't make any sense, but they are currently placed there.

**Comment by petrochenkov** [comment]
The "after" fields could actually go into `ResolverOutputs` directly, without any additional structure.
So `Definitions` will exist solely during early compilation as a part of `Resolver`.

**Comment by marmeladema** [comment]
My idea was to move everything `NodeId` related (ie: the "before" fields)  into the `Resolver` such that they are dropped with it but I would keep the rest of the fields in `Definitions`.

So to summarize, using this PR as base of discussion, `Definitions` would only contains:
* `def_id_to_hir_id`
* `hir_id_to_def_id`
* `table`
* `parent_modules_of_macro_defs`
* `expansions_that_defined`
which is the list of fields that should be available post HIR lowering.

TL;DR: "before" fields in `Resolver`, "after" fields kept in `Definitions`.

File: src/librustc_hir/hir.rs [hunk 1]
```diff
@@ -638,6 +638,8 @@ pub struct Crate<'hir> {
     /// A list of proc macro HirIds, written out in the order in which
     /// they are declared in the static array generated by proc_macro_harness.
     pub proc_macros: Vec<HirId>,
+
+    pub trait_map: BTreeMap<HirId, Vec<TraitCandidate<HirId>>>,
```
Comment (petrochenkov) [inline src/librustc_hir/hir.rs:642]: To clarify, the logic is that `Resolver::into_outputs` cannot lower `NodeId`s to `HirId`s if we remove the relevant conversion maps from `Definitions`, so we have to move it to creation of this map to some place where these conversion maps are still available, i.e. AST lowering.

Comment (petrochenkov) [reply src/librustc_hir/hir.rs:642]: `trait_map` can use `TraitCandidate<LocalDefId>` both in resolver and later.
Or rather `TraitCandidate` can be done non-generic and use `LocalDefId` directly.

Comment (petrochenkov) [reply src/librustc_hir/hir.rs:642]: Could you remove the `type TraitMap` alias as well?

File: src/librustc_middle/hir/map/collector.rs [hunk 2]
```diff
@@ -232,38 +232,13 @@ impl<'a, 'hir> NodeCollector<'a, 'hir> {
...
-                     current_dep_node_owner={} ({:?}), hir_id.owner={} ({:?})",
-                    self.source_map.span_to_string(span),
-                    node_str,
-                    self.definitions.def_path(self.current_dep_node_owner).to_string_no_crate(),
-                    self.current_dep_node_owner,
-                    self.definitions.def_path(hir_id.owner).to_string_no_crate(),
-                    hir_id.owner,
-                )
-            }
-        }
```
Comment (petrochenkov) [inline src/librustc_middle/hir/map/collector.rs:265]: Is it possible to keep this check by moving it to AST lowering, for example?
IIRC, it caught a number of issues in the past.

File: src/librustc_hir/definitions.rs [hunk 3]
```diff
@@ -461,14 +456,16 @@ impl Definitions {
...
-            "trying to initialize `NodeId` -> `HirId` mapping twice"
+            self.def_id_to_hir_id.is_empty(),
+            "trying to initialize `LocalDefId` <-> `HirId` mappings twice"
         );
-        self.node_id_to_hir_id = mapping;
 
-        // Build the pre-computed mapping of `hir_id_to_node_id` -> `node_id_to_def_id`.
-        self.hir_id_to_def_id = self
-            .node_id_to_hir_id
+        for &node_id in self.def_id_to_node_id.iter() {
```
Comment (petrochenkov) [inline src/librustc_hir/definitions.rs:463]: Nit: why doesn't it use `collect`?

**Comment by petrochenkov** [comment]
@marmeladema 
>TL;DR: "before" fields in `Resolver`, "after" fields kept in `Definitions`.

Sounds like a plan!

File: src/librustc_hir/hir.rs (hunk 1 above)
Comment (marmeladema) [reply src/librustc_hir/hir.rs:642]: So `trait_map` cannot use `LocalDefId`, it really is storing `HirId` that don't have associated `LocalDefId`, at least either for the key or the value (can't remember which, maybe both).

File: src/librustc_middle/hir/map/collector.rs (hunk 2 above)
Comment (marmeladema) [reply src/librustc_middle/hir/map/collector.rs:265]: I'll take a look

File: src/librustc_hir/definitions.rs (hunk 3 above)
Comment (marmeladema) [reply src/librustc_hir/definitions.rs:463]: For no reason actually. I'll make the change.

Comment (marmeladema) [reply src/librustc_hir/definitions.rs]: Fixed

File: src/librustc_middle/hir/map/collector.rs (hunk 2 above)
Comment (marmeladema) [reply src/librustc_middle/hir/map/collector.rs]: I wrongly removed it because i did not realized that the conversion `HirId` -> `NodeId` -> `LocalDefId` is actually the same conversion as `HirId` -> `LocalDefId` with `opt_hir_id_to_local_def_id`.

File: src/librustc_hir/hir.rs (hunk 1 above)
Comment (marmeladema) [reply src/librustc_hir/hir.rs:642]: @petrochenkov if you have any advises on how to use `LocalDefId` in the `trait_map` i am all hear because i am a bit stuck here.

Comment (petrochenkov) [reply src/librustc_hir/hir.rs:642]: https://github.com/rust-lang/rust/pull/73357
I'm sorry.

Comment (petrochenkov) [reply src/librustc_hir/hir.rs:642]: I tried to check whether are any problems with converting these in practice and there was none, so I just sent a PR.

Comment (marmeladema) [reply src/librustc_hir/hir.rs:642]: I am not sure I am following. In the PR you just sent, you are still calling `node_id_to_hir_id` to convert the keys whereas that function would be removed alongside the backing conversion table with this PR.

Comment (petrochenkov) [reply src/librustc_hir/hir.rs:642]: >`trait_map` can use `TraitCandidate<LocalDefId>` both in resolver and later.

^ The original suggestion.
It's about values of the trait map, not keys.

Keys still have to be fine-grained and use `NodeId`s early and `HirId`s later, so for keys this PR does the right thing.

**Comment by petrochenkov** [comment]
ID memo:
- `LocalDefId` - the best variant, if you can use it for something do it. Cannot address entities in other crates and fine-grained entities like individual expressions in the current crate. Can address any items.
- `DefId` - the second choice. Can address entities in other crates, but cannot address fine-grained entities like individual expressions.
- `HirId` - Can address fine-grained entities like individual expressions in the current crate. Fine-grained entities from other crates can never be addressed in rustc.
- `NodeId` - The last choice. Can address fine-grained entities like individual expressions when `HirId`s are not yet created.

File: src/librustc_hir/hir.rs (hunk 1 above)
Comment (marmeladema) [reply src/librustc_hir/hir.rs:642]: Ok sorry for the misunderstanding. I am happy to land your PR first and rebase this one on it.

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/73357 has landed.

**Comment by marmeladema** [comment]
@petrochenkov rebased and fixed the conflicts :+1:
